Avahi dependency support

daminetreg opened the pull request:
Dear @ruslo,

I find what you are doing with Hunter pretty cool. I integrated the Avahi linux-only library dependency into hunter, as this allows to use the dns_sd.h library from Apple as well on linux.

To support Avahi, I has to add 2 dependencies : libdaemon and intltool which again are really linux specific.

Have you any interests in merging this ? I'm open for improvements of my changes. 

I would like also to discuss, if it would be possible to develop an extension mechanism, to handle packages which are not yet supported by hunter, in that one could have a hunter_config/ folder and overrides/additional packages as the hunter/cmake folder. 

This would ease integration of non supported library, as it wouldn't be required to patch hunter for this. I could help in the implementation of such a solution.
 
Cheers

ruslo commented on cmake/schemes/url_sha1_avahi_autotools.cmake.in:
```diff
@@ -0,0 +1,64 @@
+     --disable-mono
+     --disable-monodoc
+     --with-xml=expat
+     --enable-compat-libdns_sd
```
I guess we need an extra flags here to pass a custom location of dependencies `Expat`, `libdaemon` and `intltool`. I've got next error while trying to build example:
```
cd /.../Hunter/_Base/93af25f/fc4ba94/b270781/Build/Avahi/Source && . /.../Hunter/_Base/Download/Hunter/unknown/93af25f/Unpacked/scripts/clear-all.sh && AR=/usr/bin/ar ./configure AR=/usr/bin/ar LD=/usr/bin/ld NM=/usr/bin/nm OBJCOPY=/usr/bin/objcopy OBJDUMP=/usr/bin/objdump RANLIB=/usr/bin/ranlib STRIP=/usr/bin/strip CC=/usr/bin/gcc CXX=/usr/bin/g++ CXXFLAGS=-std=c++11 --disable-qt3 --disable-qt4 --disable-gtk --disable-gtk3 --disable-gdbm --disable-python --disable-mono --disable-monodoc --with-xml=expat --enable-compat-libdns_sd --enable-shared --disable-static --prefix=/.../Hunter/_Base/93af25f/fc4ba94/b270781/Install
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
...
checking whether NLS is requested... yes
checking for intltool >= 0.35.0... ./configure: line 18800: intltool-update: command not found
```

daminetreg replied:
Sorry I thought I tested with intltool uninstalled, and sadly didn't. It looks like I should modify the $env{PATH} variable to the local sysroot, it would be smart to do it in toolchain.cmake.

daminetreg replied:
So in fact I've added the forwarding of the PATH variable within the hunter_autotools_project.cmake which allows the built intltool to be found. 

This also extends it for any autotools project, perhaps should I do the same for the hunter_cmake_project.cmake ?

What is however problematic with the forwarding of PATH as I did, is by cross-compiling one would like the intltool-native used to configure an avahi for another target, and this isn't handled. Actually I'm unsure if you have the concept in hunter of a sysroot-native and sysroot-target, wouldn't it make sense ?

For the search of libraries via pkgconfig with autoreconf, I also added to the configure call the PKG_CONFIG_PATH so that libraries get searched in HUNTER_PACKAGE_INSTALL_PREFIX.

ruslo replied:
> it would be smart to do it in toolchain.cmake

I think no, toolchain is a "global" object to all of the packages. It means that same toolchain can be used for two different versions of `intltool`. Since versions is different they have a different locations: `<root-id-1>/Install/bin/intltool` (v1), `<root-id-2>/Install/bin/intltool` (v2).

> So in fact I've added the forwarding of the PATH variable within the hunter_autotools_project.cmake which allows the built intltool to be found

> What is however problematic with the forwarding of PATH as I did, is by cross-compiling one would like the intltool-native used to configure an avahi for another target, and this isn't handled. 

The best choice is to add a location explicitly (like `--intltool-path=/path/to/intltool`) if there is no such option in `configure` then we need an extra argument to the `hunter_autotools_project` (like `ENV_PATH`).

> Actually I'm unsure if you have the concept in hunter of a sysroot-native and sysroot-target, wouldn't it make sense ?

There is no any "special" concept in Hunter, it just inherit CMake concept of cross-compiling. I.e. you can use `CMAKE_HOST_SYSTEM_NAME` to detect Linux host platform or `ANDROID` and `IOS` to check you're builing Android or iOS targets.

> For the search of libraries via pkgconfig with autoreconf, I also added to the configure call the PKG_CONFIG_PATH so that libraries get searched in HUNTER_PACKAGE_INSTALL_PREFIX

Probably something similar can be applied to the PATH, what do you think?

daminetreg replied:
> The best choice is to add a location explicitly (like --intltool-path=/path/to/intltool) if there is no such option in configure then we need an extra argument to the hunter_autotools_project (like ENV_PATH).

Sadly it's not possible to specify this to configure, it simply searches on PATH. That's why I added it like PKG_CONFIG_PATH : https://github.com/daminetreg/hunter/commit/49a07d7c6e936d669c0b72134edeccfd541ddce7

ruslo replied:
Just tested `ExternalProject_Add`, you can do the next:
```cmake
set(cmd "")
if(need_to_add_path) # or always?
  list(APPEND cmd PATH=${some_extra_path})
endif()
if(need_to_add_pkg_config) # or always?
  list(APPEND cmd PKG_CONFIG_PATH=${some_extra_pkg_path})
endif()
list(APPEND cmd "${path_to_configure}/configure")

ExternalProject_Add(... CONFIGURE_COMMAND ${cmd} ...)
```

ruslo commented on cmake/find/FindAvahi.cmake:
```diff
@@ -0,0 +1,69 @@
+
+if(Avahi_FOUND)
+  set(Avahi_LIBRARIES ${Avahi_COMMON_LIBRARY} ${Avahi_CLIENT_LIBRARY} ${Avahi_COMPAT_LIBDNS_SD_LIBRARY})
+  set(Avahi_INCLUDE_DIRS ${Avahi_CLIENT_INCLUDE_DIRS} ${Avahi_COMPAT_LIBDNS_SD_INCLUDE_DIRS})
```
You can create modern-style imported targets instead of old-style variables with full paths to make this libs cacheable and relocatable (though they need to be relocatable by design, also all dependencies should be cacheable too). See [this overview](https://github.com/ruslo/hunter/wiki/example.find_package.config#relocation-challenge) and [boost example](https://github.com/ruslo/hunter/blob/master/cmake/templates/BoostConfig.cmake.in). But note that it's not an issue, it's an optimization note. So you can update this part later.

daminetreg replied:
I'll change this, I'm aware of package-config file for cmake projects, but didn't know we can do it in a modern way for non-cmake projects.

daminetreg replied:
I changed this in e8cdcde to add imported targets, I'm not 100% sure of the correctness of my use of imported targets, so I stay tuned for your feedback :smile: .

ruslo replied:
Thanks, looks good for me.

ruslo commented on cmake/configs/default.cmake:
```diff
@@ -35,6 +35,7 @@ hunter_config(Android-SDK-Platform VERSION 21_r02) # API 21
 hunter_config(Android-SDK-Platform-tools VERSION r22)
 hunter_config(Android-SDK-Tools VERSION 24.3.3)
 hunter_config(Android-Support-Repository VERSION 16)
+hunter_config(Avahi VERSION 0.6.31 CMAKE_ARGS BUILD_SHARED_LIBS=ON)
```
You can use `hunter_cmake_args` in `hunter.cmake` module of the Avahi package so we can keep config file clean. See [wiki](https://github.com/ruslo/hunter/wiki/usr.adding.new.package#cmake-options) and [example](https://github.com/ruslo/hunter/blob/master/cmake/projects/Comet/hunter.cmake).

ruslo commented:
> Have you any interests in merging this ?

Sure.

> I would like also to discuss, if it would be possible to develop an extension mechanism, to handle packages which are not yet supported by hunter, in that one could have a hunter_config/ folder and overrides/additional packages as the hunter/cmake folder.

> This would ease integration of non supported library, as it wouldn't be required to patch hunter for this. I could help in the implementation of such a solution.

I was thinking about it before: https://github.com/ruslo/hunter/issues/56
I guess version with `URL` and `SHA1` is doable and quite handy for testing but note that in some cases like `Avahi` build you still need to patch Hunter since we need extra CMake modules and custom scheme.

daminetreg commented:
> I was thinking about it before: #56
I guess version with URL and SHA1 is doable and quite handy for testing but note that in some cases like Avahi build you still need to patch Hunter since we need extra CMake modules and custom scheme.

I think it's good to support well first-class citizen: that is CMake based packages, and not the others. :smile: 

If you need any help to work on #56 I'm available.

ruslo commented:
I've got this error now:
```
checking for DBUS... no
configure: error: Package requirements ( dbus-1 >= 0.34 ) were not met:

No package 'dbus-1' found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables DBUS_CFLAGS
and DBUS_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
```

daminetreg commented:
Oups... Which distribution do you use ? Because I sadly have all those dependencies installed and I don't know it.

ruslo commented:
> Oups... Which distribution do you use ? Because I sadly have all those dependencies installed and I don't know it. 

I'm testing it on Ubuntu 14.04

daminetreg commented:
So we are on the same distro, but it looks like I had dbus and intltoolize already. I'll look into adding dbus as dependency.

ruslo commented:
I've installed it system-wide using `sudo apt-get -y install libdbus-1-dev`. Let's use it as a workaround for now. I've openned this bug in case you want to add dbus too: https://github.com/ruslo/hunter/issues/228

ruslo commented on examples/Avahi/main.cpp:
```diff
@@ -0,0 +1,5 @@
+#include <avahi-client/client.h>
+#include <dns_sd.h>
```
Got this error:
```
/.../hunter/examples/Avahi/main.cpp:2:20: fatal error: dns_sd.h: No such file or directory
 #include <dns_sd.h>
                    ^
```
guess it should be (?):
```cpp
#include <avahi-client/client.h>
#include <avahi-compat-libdns_sd/dns_sd.h>

int main(int argc, char** argv) {
}
```

daminetreg replied:
That's right. That's again the mix that my system provides, as the ubuntu-avahi-compat-libdns_sd installs it directly in usr/include/. I push the change.

daminetreg commented:
Thanks.

ruslo commented:
So to summarize:
* Patch looks good for me except of the PATH and PKG_CONFIG_PATH. I've made an update of the [hunter_autotools_project](https://github.com/ruslo/hunter/commit/e3baf868061718359bf3922fcd6a8df8d5e67d7c) function. Scheme usage:
```cmake
GLOBAL_INSTALL_DIR
  "@HUNTER_INSTALL_PREFIX@"
PARALLEL_JOBS
  "@HUNTER_JOBS_OPTION@"
CPPFLAGS
  "-I@HUNTER_INSTALL_PREFIX@/include" # find Expat
LDFLAGS
  "-L@HUNTER_INSTALL_PREFIX@/lib" # find Expat
EXTRA_FLAGS
 --disable-qt3
 --disable-qt4
 --disable-gtk
 --disable-gtk3
 --disable-gdbm
 --disable-python
 --disable-mono
 --disable-monodoc
 --with-xml=expat
 --enable-compat-libdns_sd
MODIFY_PATH # find intltool-update executable (intltool package)
MODIFY_PKG_CONFIG # find libdaemon
```
* Please rebase changes against branch `develop` (see [contributing](https://github.com/ruslo/hunter#contribution))
* I've added an implementation notes to the issue #56 in case you want to experiment
* `I changed this in e8cdcde to add imported targets` now we are ready to test if can it be relocatable or not. Since it might be quite tricky for autotools project I've openned new issue for discussion: https://github.com/ruslo/hunter/issues/231

daminetreg commented:
Thank you for helping me getting it right :smile:.

I rebased on top of develop, removed the commits which were dealing with PKG_CONFIG_PATH and PATH in favor of yours and took your configuration of hunter_autotools_project.

I reopened a PR targetting develop with dbus supported also : #233
